src/metrics/metrics.ts

```diff
@@ -0,0 +1,214 @@
+import type {LogSink} from '@rocicorp/logger';
+
+export interface ReporterOptions {
+  datadogApiKey: string;
+  metrics: Metrics;
+  // intervalSecs defaults to 2 minutes.
+  intervalSecs?: number;
+  abortSignal?: AbortSignal;
+  // If a LogSink is not provided, the Reporter is silent.
+  logSink?: LogSink;
+  // Should probably also take tags (eg, env, version). Will
+  // probalby need to take host and service too for the server side.
+}
+
+// Reporter periodically reports metrics to Datadog. It uses an interval
+// instead of a timer because we want to keep to the desired interval as
+// closely as possible. A typical pattern for sampling a client is for the
+// client to report metrics on a given interval and then rollup metrics on
+// the server on the same interval. The extent to which the interval drifts
+// skews the counts seen by the server.
+export class Reporter {
+  private _metrics: Metrics;
+  private _apiKey: string;
+  private _intervalSecs: number;
+  private _intervalID: ReturnType<typeof setInterval> | 0 = 0;
+  private _abortSignal: AbortSignal | undefined;
+  private _logSink: LogSink | undefined;
+
+  constructor(options: ReporterOptions) {
+    this._metrics = options.metrics;
+    this._apiKey = options.datadogApiKey;
+    this._intervalSecs = options.intervalSecs || 2 * 60;
+    this._abortSignal = options.abortSignal;
+    this._logSink = options.logSink;
```

Comment:
We generally use an OptionalLogger instance instead.

Reply:
Can you help me understand? Require the customer to pass in an optional LogSink and if it is defined then wrap it in an optionallogger, and if it is not wrap a nooplogger in an optionalogger?

Reply:
One `OptionalLogger` that we use all over is `LogContext`. `LogSink` is a bit less ergonomic because you have to call `log('info', msg()` instead of `info(msg)`

If you use an `OptionalLogger?` instead of a `LogSink?` you do not really need the _maybeLog* methods.

Reply:
ok done